File: lab/01/src/matmul.cpp
#include <chrono>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <random>
#include <stdexcept>
#include <string>
#include <vector>

#ifdef USE_MKL
#include <mkl_cblas.h>
#endif

struct Config {
    int m = 512;
    int n = 512;
    int k = 512;
    std::string version = "naive";  // naive | reorder | unroll | mkl
    bool full_print = false;
    bool no_print = false;
};

static bool in_range(int x) {
    return x >= 512 && x <= 2048;
}

static void fill_random(std::vector<double>& v, uint32_t seed = 42) {
    std::mt19937 gen(seed);
    std::uniform_real_distribution<double> dist(-1.0, 1.0);
    for (double& x : v) x = dist(gen);
}

static void print_matrix(const std::vector<double>& mat, int rows, int cols, const std::string& name, bool full_print) {
    std::cout << name << " (" << rows << "x" << cols << "):\n";
    int r_lim = full_print ? rows : std::min(rows, 8);
    int c_lim = full_print ? cols : std::min(cols, 8);

    for (int i = 0; i < r_lim; ++i) {
        for (int j = 0; j < c_lim; ++j) {
            std::cout << std::fixed << std::setprecision(4) << mat[static_cast<size_t>(i) * cols + j] << " ";
        }
        if (!full_print && c_lim < cols) std::cout << "...";
        std::cout << "\n";
    }
    if (!full_print && r_lim < rows) {
        std::cout << "...\n";
    }
}

static void matmul_naive(const std::vector<double>& A, const std::vector<double>& B, std::vector<double>& C, int m, int n, int k) {
    std::fill(C.begin(), C.end(), 0.0);
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < k; ++j) {
            double sum = 0.0;
            for (int p = 0; p < n; ++p) {
                sum += A[static_cast<size_t>(i) * n + p] * B[static_cast<size_t>(p) * k + j];
            }
            C[static_cast<size_t>(i) * k + j] = sum;
        }
    }
}

static void matmul_reorder(const std::vector<double>& A, const std::vector<double>& B, std::vector<double>& C, int m, int n, int k) {
    std::fill(C.begin(), C.end(), 0.0);
    for (int i = 0; i < m; ++i) {
        for (int p = 0; p < n; ++p) {
            const double a = A[static_cast<size_t>(i) * n + p];
            const size_t b_base = static_cast<size_t>(p) * k;
            const size_t c_base = static_cast<size_t>(i) * k;
            for (int j = 0; j < k; ++j) {
                C[c_base + j] += a * B[b_base + j];
            }
        }
    }
}

static void matmul_unroll4(const std::vector<double>& A, const std::vector<double>& B, std::vector<double>& C, int m, int n, int k) {
    std::fill(C.begin(), C.end(), 0.0);
    for (int i = 0; i < m; ++i) {
        const size_t c_base = static_cast<size_t>(i) * k;
        for (int p = 0; p < n; ++p) {
            const double a = A[static_cast<size_t>(i) * n + p];
            const size_t b_base = static_cast<size_t>(p) * k;

            int j = 0;
            for (; j + 3 < k; j += 4) {
                C[c_base + j] += a * B[b_base + j];
                C[c_base + j + 1] += a * B[b_base + j + 1];
                C[c_base + j + 2] += a * B[b_base + j + 2];
                C[c_base + j + 3] += a * B[b_base + j + 3];
            }
            for (; j < k; ++j) {
                C[c_base + j] += a * B[b_base + j];
            }
        }
    }
}

static void matmul_mkl(const std::vector<double>& A, const std::vector<double>& B, std::vector<double>& C, int m, int n, int k) {
#ifdef USE_MKL
    cblas_dgemm(CblasRowMajor, CblasNoTrans, CblasNoTrans, m, k, n, 1.0, A.data(), n, B.data(), k, 0.0, C.data(), k);
#else
    (void)A;
    (void)B;
    (void)C;
    (void)m;
    (void)n;
    (void)k;
    throw std::runtime_error("MKL support not enabled. Rebuild with -DUSE_MKL and MKL link flags.");
#endif
}

static Config parse_args(int argc, char** argv) {
    if (argc < 5) {
        throw std::runtime_error("Usage: ./matmul <m> <n> <k> <version: naive|reorder|unroll|mkl> [--full-print] [--no-print]");
    }

    Config cfg;
    cfg.m = std::stoi(argv[1]);
    cfg.n = std::stoi(argv[2]);
    cfg.k = std::stoi(argv[3]);
    cfg.version = argv[4];

    for (int i = 5; i < argc; ++i) {
        if (std::strcmp(argv[i], "--full-print") == 0) {
            cfg.full_print = true;
        } else if (std::strcmp(argv[i], "--no-print") == 0) {
            cfg.no_print = true;
        }
    }

    if (!in_range(cfg.m) || !in_range(cfg.n) || !in_range(cfg.k)) {
        throw std::runtime_error("m, n, k must be in [512, 2048].");
    }

    if (cfg.version != "naive" && cfg.version != "reorder" && cfg.version != "unroll" && cfg.version != "mkl") {
        throw std::runtime_error("version must be one of: naive, reorder, unroll, mkl");
    }

    return cfg;
}

int main(int argc, char** argv) {
    try {
        const Config cfg = parse_args(argc, argv);

        std::vector<double> A(static_cast<size_t>(cfg.m) * cfg.n);
        std::vector<double> B(static_cast<size_t>(cfg.n) * cfg.k);
        std::vector<double> C(static_cast<size_t>(cfg.m) * cfg.k);

        fill_random(A, 42);
        fill_random(B, 777);

        auto t0 = std::chrono::steady_clock::now();

        if (cfg.version == "naive") {
            matmul_naive(A, B, C, cfg.m, cfg.n, cfg.k);
        } else if (cfg.version == "reorder") {
            matmul_reorder(A, B, C, cfg.m, cfg.n, cfg.k);
        } else if (cfg.version == "unroll") {
            matmul_unroll4(A, B, C, cfg.m, cfg.n, cfg.k);
        } else {
            matmul_mkl(A, B, C, cfg.m, cfg.n, cfg.k);
        }

        auto t1 = std::chrono::steady_clock::now();
        const std::chrono::duration<double> elapsed = t1 - t0;

        if (!cfg.no_print) {
            print_matrix(A, cfg.m, cfg.n, "A", cfg.full_print);
            print_matrix(B, cfg.n, cfg.k, "B", cfg.full_print);
            print_matrix(C, cfg.m, cfg.k, "C", cfg.full_print);
        }

        const double seconds = elapsed.count();
        const double flops = 2.0 * static_cast<double>(cfg.m) * cfg.n * cfg.k;
        const double gflops = flops / seconds / 1e9;

        std::cout << "Version: " << cfg.version << "\n";
        std::cout << "Time (sec): " << std::fixed << std::setprecision(6) << seconds << "\n";
        std::cout << "GFLOPS: " << std::fixed << std::setprecision(3) << gflops << "\n";

        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}

File: lab/00/src/matmul.py
#!/usr/bin/env python3
import random
import sys
import time


def in_range(x: int) -> bool:
    return 512 <= x <= 2048


def rand_matrix(rows: int, cols: int, seed: int):
    random.seed(seed)
    return [[random.uniform(-1.0, 1.0) for _ in range(cols)] for _ in range(rows)]


def matmul_naive(a, b, m: int, n: int, k: int):
    c = [[0.0 for _ in range(k)] for _ in range(m)]
    for i in range(m):
        for j in range(k):
            s = 0.0
            for p in range(n):
                s += a[i][p] * b[p][j]
            c[i][j] = s
    return c


def print_matrix(mat, rows: int, cols: int, name: str, full_print: bool):
    print(f"{name} ({rows}x{cols}):")
    r_lim = rows if full_print else min(rows, 8)
    c_lim = cols if full_print else min(cols, 8)

    for i in range(r_lim):
        row_vals = " ".join(f"{mat[i][j]:.4f}" for j in range(c_lim))
        suffix = " ..." if (not full_print and c_lim < cols) else ""
        print(row_vals + suffix)

    if not full_print and r_lim < rows:
        print("...")


def parse_args(argv):
    if len(argv) < 4:
        raise ValueError("Usage: python3 matmul.py <m> <n> <k> [--full-print] [--no-print]")

    m = int(argv[1])
    n = int(argv[2])
    k = int(argv[3])
    full_print = "--full-print" in argv[4:]
    no_print = "--no-print" in argv[4:]

    if not in_range(m) or not in_range(n) or not in_range(k):
        raise ValueError("m, n, k must be in [512, 2048].")

    return m, n, k, full_print, no_print


def main():
    try:
        m, n, k, full_print, no_print = parse_args(sys.argv)

        a = rand_matrix(m, n, 42)
        b = rand_matrix(n, k, 777)

        t0 = time.perf_counter()
        c = matmul_naive(a, b, m, n, k)
        t1 = time.perf_counter()

        if not no_print:
            print_matrix(a, m, n, "A", full_print)
            print_matrix(b, n, k, "B", full_print)
            print_matrix(c, m, k, "C", full_print)

        sec = t1 - t0
        flops = 2.0 * m * n * k
        gflops = flops / sec / 1e9

        print("Version: python")
        print(f"Time (sec): {sec:.6f}")
        print(f"GFLOPS: {gflops:.3f}")
    except Exception as exc:
        print(f"Error: {exc}", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
